### PR TITLE
feat(ci): Balance frontend tests

### PR DESCRIPTION
**Pull request by billyvg**
This adds a jest reporter to record the test durations of our frontend tests. It also changes our runner to look for a results JSON and attempt to balance tests between our CI runners.

### CODE REVIEW

File: jest.config.ts [hunk 1]
```diff
@@ -20,25 +27,111 @@ const {CI, JEST_TESTS, CI_NODE_TOTAL, CI_NODE_INDEX, GITHUB_PR_SHA, GITHUB_PR_RE
...
+ *
+ * Returns a map of <testName, groupIndex>
+ */
+function balancer(
+  allTests: string[],
+  source: Record<string, number>,
+  numberGroups: number
+) {
+  const results = new Map<string, number>();
+  const totalRunTimes = Array.from(Array(numberGroups)).map(() => 0);
```
Comment (evanpurkhiser) [inline jest.config.ts:54]:
```suggestion
  const totalRunTimes = Array(numberGroups).fill(0);
```

Thoughts?

File: tests/js/test-balancer/index.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    ) {
+      return;
+    }
+
+    const data = JSON.stringify(Object.fromEntries(this.results), null, 2);
+    fs.writeFile(this._options.resultsPath, data);
+  }
+}
+
+module.exports = TestBalancer;
```
Comment (evanpurkhiser) [inline tests/js/test-balancer/index.js:34]: Can you write this in typescript as well?

Comment (billyvg) [reply tests/js/test-balancer/index.js:34]: Not possible unfortunately: https://github.com/facebook/jest/issues/8810

**Comment by billyvg** [comment]
Here's an example of frontend tests have wildly differing test sharding:

[frontend (0)](https://github.com/getsentry/sentry/runs/4729126405?check_suite_focus=true) taking 23 minutes
[frontend (1)](https://github.com/getsentry/sentry/runs/4729126476?check_suite_focus=true) taking 17 minutes

**Comment by billyvg** [comment]
I think this is useful enough to merge as-is - ideally we would add some automation to update the balance JSON file. But I think even without automating it, this should suffice for awhile (until we add *new* tests that are quite slow).

**Review by evanpurkhiser — APPROVED** [review]
looks great, I like this.

**Comment by billyvg** [comment]
Not sure what's up with the jest snapshots, but the tests are definitely getting run: 
image

**Comment by priscilawebdev** [comment]
I like it too 🙂 👏

**Review by priscilawebdev — APPROVED** [review]

**Review by narsaynorath — APPROVED** [review]
🤞

File: tests/js/spec/views/dashboardsV2/gridLayout/create.spec.tsx [hunk 3]
```diff
@@ -81,9 +85,8 @@ describe('Dashboards > Create', function () {
         // Wrap with act because GlobalSelectionHeaderContainer triggers update
         await tick();
       });
-      screen.getByTestId('widget-add').click();
 
-      mountGlobalModal();
+      screen.getByTestId('widget-add').click();
```
Comment (priscilawebdev) [inline tests/js/spec/views/dashboardsV2/gridLayout/create.spec.tsx:89]: should this be `userEvent.click(screen.getByTestId('widget-add')) `?

File: tests/js/spec/views/dashboardsV2/gridLayout/create.spec.tsx [hunk 4]
```diff
@@ -81,9 +85,8 @@ describe('Dashboards > Create', function () {
...
 
       // Add a custom widget to the dashboard
       (await screen.findByText('Custom Widget')).click();
```
Comment (priscilawebdev) [inline tests/js/spec/views/dashboardsV2/gridLayout/create.spec.tsx:92]: here too?

**Pull request merged**